### ISSUE TITLE
Chunking along dimensions of images

### ISSUE THREAD

**Issue by TomNicholas** [ocetrac/ocetrac#14]
The [dask-image library](https://github.com/dask/dask-image) has limited support for dask-powered versions of the `scipy.ndimage` functions that ocetrac wraps.

What is a bit confusing though is that some of the functions in dask-image are merely parallelizing across separate images (i.e. each worker is given one entire image to label out of 1000 images), while some are actually parallelizing the labelling of a single image (i.e. each worker is given a subregion within a single image, and together they manage to label the whole image).

The `ndimage.label` function is [properly parallelized](https://github.com/dask/dask-image/pull/94) in the latter way though. If dask-image provided the full API then you could basically run ocetrac on datasets bigger than your memory just by changing which library you imported :exploding_head:

(The problem Ocetrac solves is also very similar to [something I wanted to do](https://github.com/dask/dask-image/issues/29#issuecomment-455237486) at one point in my PhD, funnily enough)

I don't know if your current work is ever limited in this way, and dask-image does not provide a complete substitute (yet) but I thought it might be useful to know about if you didn't already.

**Comment by hscannell** [comment]
Thanks for linking this [issue](https://github.com/dask/dask-image/issues/29#issuecomment-455237486). @jbusecke was going to look into this further. I believe he was trying to implement `map_overlap` to get information from adjacent grid cells.

**Comment by jbusecke** [comment]
I think the key function is already using [dask_image](https://github.com/ocetrac/ocetrac/blob/c632982bf0bbae217cc02f737f772fdd5f9bb485/ocetrac/model.py#L195-L209).

I was trying to wrap everything 'leading up' to that into a wrapper and just throw map_blocks on it, but I ran into some trouble for now. I really need to understand the core algo better, I feel.

**Comment by TomNicholas** [comment]
> I think the key function is already using dask_image.

Oh I didn't spot that! There might be bottlenecks from the fact that the binary closing and opening is not dask-ified though.

**Comment by jbusecke** [comment]
> There might be bottlenecks from the fact that the binary closing and opening is not dask-ified though.

The binary stuff is already dask enabled. The tests in #16 currrently hang at the `_wrap` stage. Ill try to dig further into this in the coming days...

**Comment by TomNicholas** [comment]
> The binary stuff is already dask enabled. 

This confuses me because isn't it a direct call to `scipy.ndimage.binary_[etc]`? Wouldn't that eagerly operate on the arrays? It's definitely not going to parallelise the erosion operation internally surely? Anyway we can discuss in a hack if this comes up.

**Comment by jbusecke** [comment]
Yes, but that is wrapped in `apply_ufunc` [here](https://github.com/ocetrac/ocetrac/blob/df6aaa67b01af7983efdc53cf33e2cf03871e894/ocetrac/model.py#L144-L149). Does that clear it up?

**Comment by TomNicholas** [comment]
I think we're talking at cross purposes. It's great that these functions are wrapped by dask to operate on each image individually, but they still have non-chunkable core dims. The issues I linked to above specifically concern being able to chunk even along the dimensions of the image itself (rather than sets of different images). If I understand the dask-image work correctly then right now the label function can actually parallelize along x and y if desired but the binary_open function cannot. This would be useful for being able to find features in very large images.

Maybe that's just not a problem at all for this project, but I thought it was worth mentioning.

**Comment by jbusecke** [comment]
Ahhh, sorry for the misunderstanding. I think this might become relevant for future functionality (detecting 3d features?), but at this point, my main concern is to avoid triggering a full load of the dataset. But the point is well taken and we should keep this in mind for future improvements.

**Comment by hdrake** [comment]
Hi @hscannell (and @jbusecke @TomNicholas), thanks for this great package! @mwtip and I are trying to apply this to GFDL output and are having our kernels crash when trying to run `ocetrac.Trakcer` on even just a year of our `tos` data at native resolution (0.25º global). It seems probable that this has to do with either this issue or https://github.com/ocetrac/ocetrac/issues/10.

Any advice? Has these been any more progress on this issue? Is there any way to way `ocetrak` on a dataset that does not fit into memory?

Depending on how easy the fix is, I could try to take a crack at it.

**Comment by jbusecke** [comment]
Hi @hdrake. Unfortunately I personally do not have the bandwidth at this point to support ocetrac actively and am pretty out of the loop here. I hope @hscannell might be able to provide some more information.
